statistics/histogram.go

```diff
@@ -834,29 +875,69 @@ func (c *Column) IsInvalid(sc *stmtctx.StatementContext, collPseudo bool) bool {
 	if collPseudo && c.NotAccurate() {
 		return true
 	}
-	if c.NDV > 0 && c.Len() == 0 && sc != nil {
+	// If stats ver < 2, histogram represents all data in this column, so empty histogram implies unloaded stats.
+	// In stats ver 2, histogram + TopN represents all data in a column, so both of them being empty implies unloaded stats.
+	if c.NDV > 0 && sc != nil &&
+		((c.Len() == 0 && c.StatsVer < Version2) || (c.Len() == 0 && c.TopN.TotalCount() == 0 && c.StatsVer == Version2)) {
 		sc.SetHistogramsNotLoad()
 		HistogramNeededColumns.insert(tableColumnID{TableID: c.PhysicalID, ColumnID: c.Info.ID})
 	}
-	return c.TotalRowCount() == 0 || (c.NDV > 0 && c.Len() == 0)
+	return c.TotalRowCount() == 0 ||
+		(c.NDV > 0 &&
+			((c.Len() == 0 && c.StatsVer < Version2) || (c.Len() == 0 && c.TopN.TotalCount() == 0 && c.StatsVer == Version2)))
 }
 
 func (c *Column) equalRowCount(sc *stmtctx.StatementContext, val types.Datum, modifyCount int64) (float64, error) {
 	if val.IsNull() {
 		return float64(c.NullCount), nil
 	}
-	// All the values are null.
-	if c.Histogram.Bounds.NumRows() == 0 {
-		return 0.0, nil
+	if c.StatsVer < Version2 {
+		// All the values are null.
+		if c.Histogram.Bounds.NumRows() == 0 {
+			return 0.0, nil
+		}
+		if c.NDV > 0 && c.outOfRange(val) {
+			return outOfRangeEQSelectivity(c.NDV, modifyCount, int64(c.TotalRowCount())) * c.TotalRowCount(), nil
+		}
+		if c.CMSketch != nil {
+			count, err := queryValue(sc, c.CMSketch, c.TopN, val)
+			return float64(count), errors.Trace(err)
+		}
+		return c.Histogram.equalRowCount(val), nil
+	}
+	// Stats version == 2
+	// 1. try to find this value in TopN
+	if c.TopN != nil {
+		valBytes, err := codec.EncodeKey(sc, nil, val)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		rowcount, ok := c.QueryTopN(valBytes)
+		if ok {
+			return float64(rowcount), nil
+		}
 	}
-	if c.NDV > 0 && c.outOfRange(val) {
-		return outOfRangeEQSelectivity(c.NDV, modifyCount, int64(c.TotalRowCount())) * c.TotalRowCount(), nil
+	// 2. try to find this value in bucket.repeats(the last value in every bucket)
+	index, match := c.Histogram.Bounds.LowerBound(0, &val)
+	if index%2 == 1 && match {
+		return float64(c.Histogram.Buckets[index/2].Repeat), nil
 	}
-	if c.CMSketch != nil {
-		count, err := queryValue(sc, c.CMSketch, c.TopN, val)
-		return float64(count), errors.Trace(err)
+	if match {
+		cmp := chunk.GetCompareFunc(c.Histogram.Tp)
+		if cmp(c.Histogram.Bounds.GetRow(index), 0, c.Histogram.Bounds.GetRow(index+1), 0) == 0 {
+			return float64(c.Histogram.Buckets[index/2].Repeat), nil
+		}
 	}
-	return c.Histogram.equalRowCount(val), nil
+	// 3. use evenly distribution assumption for the rest
```

Comment:
Should we check if it is out-of-range before step3?

Reply:
We cannot guarantee the data is really out of range based on sampling. It's a dangerous assumption.

Reply:
winoros has given explanations so I'll mark this as resolved. @qw4990